client/extensions/woocommerce/woocommerce-services/views/shipping-label/test/label-item.js

```diff
@@ -0,0 +1,85 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { expect } from 'chai';
+import { shallow } from 'enzyme';
+import { merge } from 'lodash';
+import { translate } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import { LabelItem } from '../label-item.js';
+import PopoverMenuItem from 'components/popover/menu-item';
+
+function createLabelItemWrapper( props = {} ) {
+	const defaults = {
+		siteId: 1,
+		orderId: 1,
+		label: {
+			serviceName: "test",
+			labelIndex: 1,
+			packageName: "test package",
+			productNames: "test product",
+			receiptId: 1,
+			labelId: 1,
+			createdDate: "",
+			refundableAmount: 1,
+			currency: "",
+			showDetails: true,
+			expiryDate: "",
+			anonymized: false,
+			usedDate: "",
+			tracking: "tracking code 01",
+			carrierId: 1,
+		},
+		isModal: false,
+		openRefundDialog: () => {},
+		openReprintDialog: () => {},
+		openDetailsDialog: () => {},
+		translate: translate,
+	};
+
+	props = merge( defaults, props );
+
+	return shallow( <LabelItem { ...props } /> );
+}
+
+describe( 'Label item', () => {
+	describe( 'with customs info', () => {
+		const props = {
+			label: {
+				commercialInvoiceUrl: 'commercial-invoice.pdf'
+			}
+		};
+		const wrapper = createLabelItemWrapper( props );
+		const renderedPrintCustomsFormLink = wrapper.findWhere( ( n ) => {
+			return n.is( PopoverMenuItem ) && 'Print customs form' === n.children().text();
+		}  );
+
+		it( 'renders a link to print it', function () {
+			expect( renderedPrintCustomsFormLink ).to.have.lengthOf( 1 );
+
+			window.open = jest.fn();
+			renderedPrintCustomsFormLink.simulate( 'click' );
+			expect( window.open.mock.calls.length ).to.equal( 1 );
```

Comment:
(minor) What do you think about using `toHaveBeenCalledWith` for the expectation? I'm thinking that it could help with readability 🤷 

What do you think about changing the structure from
```
Label item > with customs info > renders a link to print it
Label item > without customs info > renders a link to print it
```
to
```
LabelItem (name of the component) > renders the customs form when present
LabelItem > does not render the customs form when not present
```

Something like this:
```diff
diff --git a/client/extensions/woocommerce/woocommerce-services/views/shipping-label/test/label-item.js b/client/extensions/woocommerce/woocommerce-services/views/shipping-label/test/label-item.js
index e8074201..b3746366 100644
--- a/client/extensions/woocommerce/woocommerce-services/views/shipping-label/test/label-item.js
+++ b/client/extensions/woocommerce/woocommerce-services/views/shipping-label/test/label-item.js
@@ -4,82 +4,60 @@
  * External dependencies
  */
 import React from 'react';
-import { expect } from 'chai';
 import { shallow } from 'enzyme';
-import { merge } from 'lodash';
 import { translate } from 'i18n-calypso';
 
 /**
  * Internal dependencies
  */
 import { LabelItem } from '../label-item.js';
-import PopoverMenuItem from 'components/popover/menu-item';
 
-function createLabelItemWrapper( props = {} ) {
-	const defaults = {
-		siteId: 1,
-		orderId: 1,
-		label: {
-			serviceName: "test",
-			labelIndex: 1,
-			packageName: "test package",
-			productNames: "test product",
-			receiptId: 1,
-			labelId: 1,
-			createdDate: "",
-			refundableAmount: 1,
-			currency: "",
-			showDetails: true,
-			expiryDate: "",
-			anonymized: false,
-			usedDate: "",
-			tracking: "tracking code 01",
-			carrierId: 1,
-		},
-		isModal: false,
-		openRefundDialog: () => {},
-		openReprintDialog: () => {},
-		openDetailsDialog: () => {},
-		translate: translate,
-	};
+const defaultProps = {
+	siteId: 1,
+	orderId: 1,
+	label: {
+		serviceName: "test",
+		labelIndex: 1,
+		packageName: "test package",
+		productNames: "test product",
+		receiptId: 1,
+		labelId: 1,
+		createdDate: "",
+		refundableAmount: 1,
+		currency: "",
+		showDetails: true,
+		expiryDate: "",
+		anonymized: false,
+		usedDate: "",
+		tracking: "tracking code 01",
+		carrierId: 1,
+	},
+	isModal: false,
+	openRefundDialog: () => {},
+	openReprintDialog: () => {},
+	openDetailsDialog: () => {},
+	translate,
+};
 
-	props = merge( defaults, props );
+describe( 'LabelItem', () => {
+	describe( 'renders the customs form when present', () => {
+		window.open = jest.fn();
+		const wrapper = shallow(<LabelItem {...defaultProps} label={{...defaultProps.label, commercialInvoiceUrl: 'customs-form.pdf'}} />)
 
-	return shallow( <LabelItem { ...props } /> );
-}
+		const customFormsLink = wrapper.find('[children="Print customs form"]');
 
-describe( 'Label item', () => {
-	describe( 'with customs info', () => {
-		const props = {
-			label: {
-				commercialInvoiceUrl: 'commercial-invoice.pdf'
-			}
-		};
-		const wrapper = createLabelItemWrapper( props );
-		const renderedPrintCustomsFormLink = wrapper.findWhere( ( n ) => {
-			return n.is( PopoverMenuItem ) && 'Print customs form' === n.children().text();
-		}  );
+		expect(customFormsLink.length).toBe(1)
 
-		it( 'renders a link to print it', function () {
-			expect( renderedPrintCustomsFormLink ).to.have.lengthOf( 1 );
-
-			window.open = jest.fn();
-			renderedPrintCustomsFormLink.simulate( 'click' );
-			expect( window.open.mock.calls.length ).to.equal( 1 );
-			expect( window.open.mock.calls[0][0] ).to.equal( 'commercial-invoice.pdf' );
-		} );
+		customFormsLink.simulate( 'click' );
 
+		expect( window.open ).toHaveBeenCalledWith( 'customs-form.pdf' );
 	} );
 
-	describe( 'without customs info', () => {
-		const wrapper = createLabelItemWrapper();
-		const renderedPrintCustomsFormLink = wrapper.findWhere( ( n ) => {
-			return n.is( PopoverMenuItem ) && 'Print customs form' === n.children().text();
-		}  );
+	describe( 'does not render the customs form when not present', () => {
+		const wrapper = shallow(<LabelItem {...defaultProps} />)
 
-		it( 'renders a link to print it', function () {
-			expect( renderedPrintCustomsFormLink ).to.have.lengthOf( 0 );
-		} );
+		const customFormsLink = wrapper.find('[children="Print customs form"]');
 
+		expect(customFormsLink.length).toBe(0)
 	} );
 } );
```

I also changed the `wrapper.findWhere` to just `wrapper.find('[children="Print customs form"]')`.
I like how you ensured the test was finding an element by its text 👍  (like React Testing Library would do)

Reply:
> (minor) What do you think about using `toHaveBeenCalledWith` for the expectation? I'm thinking that it could help with readability shrug

Totally! But I tried using that method and apparently our chai version does not support it.

 
> What do you think about changing the structure from
> 
> ```
> Label item > with customs info > renders a link to print it
> Label item > without customs info > renders a link to print it
> ```
> 
> to
> 
> ```
> LabelItem (name of the component) > renders the customs form when present
> LabelItem > does not render the customs form when not present

I'm totally fine with any structure. I've been using this one for several years now after I learnt it from a coworker who was super into BDD. The idea behind it is you define:

what do you want to test > context > behaviour this context enables

IMHO this groups tests better and make them easier to read in the final test output.

But that's totally my point of view. Maybe we should talk about it in our next dev roundtable? 

> I also changed the `wrapper.findWhere` to just `wrapper.find('[children="Print customs form"]')`.
> I like how you ensured the test was finding an element by its text +1 (like React Testing Library would do)

I'm not super familiar with Jest syntax (actually, took me a while to come up with that `findWhere` filter), but what I was trying to accomplish here was: find a `PopoverMenuItem` with text `'Print customs form'`. Not sure if there's another way to do this ensuring both conditions. What are your thoughts on this?

Reply:
> Totally! But I tried using that method and apparently our chai version does not support it.

Yeah, I removed the `import { expect } from 'chai`;`, the `expect` in my diff is coming from `jest`, I believe 👍 

> [...] Maybe we should talk about it in our next dev roundtable?

Sure thing! And again, it's minor to me :) what you have is totes good, I appreciate the explanation 👍 

> I'm not super familiar with Jest syntax (actually, took me a while to come up with that findWhere filter), but what I was trying to accomplish here was: find a PopoverMenuItem with text 'Print customs form'. Not sure if there's another way to do this ensuring both conditions. What are your thoughts on this?

I believe you could do `wrapper.find('PopoverMenuItem[children="Print customs form"]')`.
However, this is testing the implementation details of a component (in RTL for example, there isn't an API that exposes components by their names for this reason - they don't expose component names by design)

Reply:
> Yeah, I removed the `import { expect } from 'chai`;`, the `expect`in my diff is coming from`jest`, I believe +1

Ok, I'll try that :+1: 

> Sure thing! And again, it's minor to me :) what you have is totes good, I appreciate the explanation +1

:+1: 
> I believe you could do `wrapper.find('PopoverMenuItem[children="Print customs form"]')`.

I'll try that.

> However, this is testing the implementation details of a component (in RTL for example, there isn't an API that exposes components by their names for this reason - they don't expose component names by design)

Thanks, TIL! I'm not familiar at all with Javascript component testing. Why is not a good practice to test the implementation details? Doesn't that create more fragile tests?

Reply:
> Thanks, TIL! I'm not familiar at all with Javascript component testing. Why is not a good practice to test the implementation details? Doesn't that create more fragile tests?

There is definitely some give and take on the different approaches. But it seems that, when testing React components that provide UI to the user, it's better to test how the user interacts with the UI. This ensures that tests will be breaking less often when refactoring.
I found [this article](https://kentcdodds.com/blog/testing-implementation-details) to be an ok read in the past - a good starting point, at least :)

Reply:
Ok, thanks @frosso ! I'll take a look at that article, really appreciated!